libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationService.java

```diff
@@ -244,51 +252,63 @@ public void onMilestoneTrigger(List<Milestone> triggeredMilestones, RouteProgres
    * whether or not the user is off route, the event dispatcher gets called.
    */
   @Override
   public void onUserOffRoute(Location location, boolean userOffRoute) {
     if (userOffRoute) {
       if (location.getTime() > timeIntervalSinceLastOffRoute
         + TimeUnit.SECONDS.toMillis(mapboxNavigation.options().secondsBeforeReroute())) {
-        recentDistancesFromManeuverInMeters.clear();
-        mapboxNavigation.getEventDispatcher().onUserOffRoute(location);
         timeIntervalSinceLastOffRoute = location.getTime();
+        if (mapboxNavigation.getSessionState().lastReroutePosition() == null) {
+          rerouteSessionsStateUpdate();
+        } else {
+          if (TurfMeasurement.distance(mapboxNavigation.getSessionState().lastReroutePosition(),
+            Position.fromLngLat(location.getLongitude(), location.getLatitude()),
+            TurfConstants.UNIT_METERS)
+            > mapboxNavigation.options().minimumDistanceBeforeRerouting()) {
+            mapboxNavigation.getEventDispatcher().onUserOffRoute(location);
+            rerouteSessionsStateUpdate();
+          }
+        }
       }
     } else {
       timeIntervalSinceLastOffRoute = location.getTime();
     }
   }
 
+  private void rerouteSessionsStateUpdate() {
+    recentDistancesFromManeuverInMeters.clear();
+    mapboxNavigation.getEventDispatcher().onUserOffRoute(rawLocation);
+    mapboxNavigation.setSessionState(
+      mapboxNavigation.getSessionState().toBuilder().lastReroutePosition(
+        Position.fromLngLat(rawLocation.getLongitude(), rawLocation.getLatitude())).build()
+    );
+  }
+
   public void rerouteOccurred() {
     mapboxNavigation.setSessionState(mapboxNavigation.getSessionState().toBuilder()
+      .routeProgressBeforeReroute(routeProgress)
       .beforeRerouteLocations(Arrays.asList(
         locationBuffer.toArray(new Location[locationBuffer.size()])))
-      .routeProgressBeforeReroute(routeProgress)
+      .lastRerouteDate(new Date())
       .build());
-    locationBuffer.clear();
-    queuedRerouteEvent = true;
-
-    handler = new Handler();
-    runnable = new Runnable() {
-      @Override
-      public void run() {
-        mapboxNavigation.setSessionState(mapboxNavigation.getSessionState().toBuilder()
-          .afterRerouteLocations(Arrays.asList(
-            locationBuffer.toArray(new Location[locationBuffer.size()])))
-          .build());
-
-        locationBuffer.clear();
-
-        NavigationMetricsWrapper.rerouteEvent(mapboxNavigation.getSessionState(), routeProgress,
-          location);
-        queuedRerouteEvent = false;
-      }
-    };
-    handler.postDelayed(runnable, TWENTY_SECOND_INTERVAL);
+
+    queuedRerouteEvents.add(mapboxNavigation.getSessionState());
+  }
+
+  void sendRerouteEvent(SessionState sessionState) {
+    mapboxNavigation.setSessionState(sessionState.toBuilder()
+      .afterRerouteLocations(Arrays.asList(
+        locationBuffer.toArray(new Location[locationBuffer.size()])))
```

Comment:
This assumes locations are always exactly 1s apart. There's no guarantee that all of the locations in the locationBuffer will have been from after the original reroute timestamp. If locations come in less frequently due to poor GPS reception, this will include locations from before the reroute. Likewise, if locations come in more frequently than 1 per second, the first few locations after the reroute will have been dropped from the buffer.